src/phoenix/core/traces.py

```diff
@@ -105,42 +106,39 @@ def __hash__(self) -> int:
         return id(self)
 
 
-ParentSpanID: TypeAlias = SpanID
-ChildSpanID: TypeAlias = SpanID
+_ParentSpanID: TypeAlias = SpanID
+_ChildSpanID: TypeAlias = SpanID
+_ProjectName: TypeAlias = str
+
+_DEFAULT_PROJECT_NAME: str = "default"
+
+_QueueItem = Optional[Union[otlp.Span, Tuple[otlp.Span, Optional[_ProjectName]]]]
 
 
 class Traces:
     def __init__(self) -> None:
-        self._queue: "SimpleQueue[Optional[otlp.Span]]" = SimpleQueue()
+        self._queue: "SimpleQueue[_QueueItem]" = SimpleQueue()
         # Putting `None` as the sentinel value for queue termination.
         weakref.finalize(self, self._queue.put, END_OF_QUEUE)
         self._lock = RLock()
-        self._spans: Dict[SpanID, ReadableSpan] = {}
-        self._parent_span_ids: Dict[SpanID, ParentSpanID] = {}
-        self._traces: DefaultDict[TraceID, Set[ReadableSpan]] = defaultdict(set)
-        self._child_spans: DefaultDict[SpanID, Set[ReadableSpan]] = defaultdict(set)
-        self._num_documents: DefaultDict[SpanID, int] = defaultdict(int)
-        self._start_time_sorted_spans: SortedKeyList[ReadableSpan] = SortedKeyList(
-            key=lambda span: span.start_time,
-        )
-        self._start_time_sorted_root_spans: SortedKeyList[ReadableSpan] = SortedKeyList(
-            key=lambda span: span.start_time,
-        )
-        self._latency_sorted_root_spans: SortedKeyList[ReadableSpan] = SortedKeyList(
-            key=lambda span: span[ComputedAttributes.LATENCY_MS.value],
-        )
-        self._root_span_latency_ms_sketch = DDSketch()
-        self._token_count_total: int = 0
-        self._last_updated_at: Optional[datetime] = None
+        self._projects: DefaultDict[_ProjectName, _Project] = defaultdict(_Project)
         self._start_consumer()
 
-    def put(self, span: Optional[otlp.Span] = None) -> None:
-        self._queue.put(span)
+    def put(self, item: _QueueItem) -> None:
+        self._queue.put(item)
 
-    def get_trace(self, trace_id: TraceID) -> Iterator[Span]:
+    def get_trace(
+        self,
+        trace_id: TraceID,
+        project_name: Optional[str] = None,
+    ) -> Iterator[Span]:
+        if not project_name:
+            project_name = _DEFAULT_PROJECT_NAME
```

Comment:
Question - why would be not just store traces per project? Is that not the ultimate goal to have some semblance of tenancy? Or is this mainly a stop gap?

Reply:
discussed offline